File: C++/Trees/red_black.cpp
#include<iostream>
using namespace std;
struct node
{
       int key;
       node *parent;
       char color;
       node *left;
       node *right;
};
class RB_tree
{
      node *root;
      node *q;
   public :
      RB_tree()
      {
              q=NULL;
              root=NULL;
      }
      void insert();
      void insert_fix(node *);
      void left_rotate(node *);
      void right_rotate(node *);
      void del();
      node* successor(node *);
      void delfix(node *);
      void disp();
      void display( node *);
      void search();
};
void RB_tree::insert()
{
     int z,i=0;
     cout<<"\nENTER THE KEY OF THE NODE TO BE INSERTED: "; cin>>z;
     node *p,*q;
     node *t=new node;
     t->key=z;
     t->left=NULL;
     t->right=NULL;
     t->color='r';
     p=root;
     q=NULL;
     if(root==NULL)
     {
           root=t;
           t->parent=NULL;
     }
     else
     {
         while(p!=NULL)
         {
              q=p;
              if(p->key<t->key)
                  p=p->right;
              else
                  p=p->left;
         }
         t->parent=q;
         if(q->key<t->key)
              q->right=t;
         else
              q->left=t;
     }
     insert_fix(t);
}
void RB_tree::insert_fix(node *t)
{
     node *u;
     if(root==t)
     {
         t->color='b';
         return;
     }
     while(t->parent!=NULL&&t->parent->color=='r')
     {
           node *g=t->parent->parent;
           if(g->left==t->parent)
           {
                        if(g->right!=NULL)
                        {
                              u=g->right;
                              if(u->color=='r')
                              {
                                   t->parent->color='b';
                                   u->color='b';
                                   g->color='r';
                                   t=g;
                              }
                        }
                        else
                        {
                            if(t->parent->right==t)
                            {
                                 t=t->parent;
                                 left_rotate(t);
                            }
                            t->parent->color='b';
                            g->color='r';
                            right_rotate(g);
                        }
           }
           else
           {
                        if(g->left!=NULL)
                        {
                             u=g->left;
                             if(u->color=='r')
                             {
                                  t->parent->color='b';
                                  u->color='b';
                                  g->color='r';
                                  t=g;
                             }
                        }
                        else
                        {
                            if(t->parent->left==t)
                            {
                                   t=t->parent;
                                   right_rotate(t);
                            }
                            t->parent->color='b';
                            g->color='r';
                            left_rotate(g);
                        }
           }
           root->color='b';
     }
}
void RB_tree::del()
{
     if(root==NULL)
     {
           cout<<"\nEMPTY TREE." ;
           return ;
     }
     int x;
     cout<<"\nENTER THE KEY OF THE NODE TO BE DELETED: "; cin>>x;
     node *p;
     p=root;
     node *y=NULL;
     node *q=NULL;
     int found=0;
     while(p!=NULL&&found==0)
     {
           if(p->key==x)
               found=1;
           if(found==0)
           {
                 if(p->key<x) p=p->right;
                 else
                    p=p->left;
           }
     }
     if(found==0)
     {
            cout<<"\nELEMENT NOT FOUND.";
            return ;
     }
     else
     {
         cout<<"\nDELETED ELEMENT: "<<p->key;
         cout<<"\nCOLOUR: "; if(p->color=='b')
     cout<<"BLACK\n";
    else
     cout<<"RED\n"; if(p->parent!=NULL)
               cout<<"\nPARENT: "<<p->parent->key;
         else
               cout<<"\nTHERE IS NO PARENT OF THE NODE. "; if(p->right!=NULL)
               cout<<"\nRIGHT CHILD "<<p->right->key;
         else
               cout<<"\nTHERE IS NO RIGHT CHILD OF THE NODE. "; if(p->left!=NULL)
               cout<<"\nLEFT CHILD: "<<p->left->key;
         else
               cout<<"\nTHERE IS NO RIGHT CHILD OF THE NODE. ";
         cout<<"\nNODE DELETED."; if(p->left==NULL||p->right==NULL)
              y=p;
         else
              y=successor(p);
         if(y->left!=NULL)
              q=y->left;
         else
         {
              if(y->right!=NULL)
                   q=y->right;
              else
                   q=NULL;
         }
         if(q!=NULL)
              q->parent=y->parent;
         if(y->parent==NULL)
              root=q;
         else
         {
             if(y==y->parent->left)
                y->parent->left=q;
             else
                y->parent->right=q;
         }
         if(y!=p)
         {
             p->color=y->color;
             p->key=y->key;
         }
         if(y->color=='b')
             delfix(q);
     }
}
void RB_tree::delfix(node *p)
{
    node *s;
    while(p!=root&&p->color=='b')
    {
          if(p->parent->left==p)
          {
                  s=p->parent->right;
                  if(s->color=='r')
                  {
                         s->color='b';
                         p->parent->color='r';
                         left_rotate(p->parent);
                         s=p->parent->right;
                  }
                  if(s->right->color=='b'&&s->left->color=='b')
                  {
                         s->color='r';
                         p=p->parent;
                  }
                  else
                  {
                      if(s->right->color=='b')
                      {
                             s->left->color=='b';
                             s->color='r';
                             right_rotate(s);
                             s=p->parent->right;
                      }
                      s->color=p->parent->color;
                      p->parent->color='b';
                      s->right->color='b';
                      left_rotate(p->parent);
                      p=root;
                  }
          }
          else
          {
                  s=p->parent->left;
                  if(s->color=='r')
                  {
                        s->color='b';
                        p->parent->color='r';
                        right_rotate(p->parent);
                        s=p->parent->left;
                  }
                  if(s->left->color=='b'&&s->right->color=='b')
                  {
                        s->color='r';
                        p=p->parent;
                  }
                  else
                  {
                        if(s->left->color=='b')
                        {
                              s->right->color='b';
                              s->color='r';
                              left_rotate(s);
                              s=p->parent->left;
                        }
                        s->color=p->parent->color;
                        p->parent->color='b';
                        s->left->color='b';
                        right_rotate(p->parent);
                        p=root;
                  }
          }
       p->color='b';
       root->color='b';
    }
}
void RB_tree::left_rotate(node *p)
{
     if(p->right==NULL)
           return ;
     else
     {
           node *y=p->right;
           if(y->left!=NULL)
           {
                  p->right=y->left;
                  y->left->parent=p;
           }
           else
                  p->right=NULL;
           if(p->parent!=NULL)
                y->parent=p->parent;
           if(p->parent==NULL)
                root=y;
           else
           {
               if(p==p->parent->left)
                       p->parent->left=y;
               else
                       p->parent->right=y;
           }
           y->left=p;
           p->parent=y;
     }
}
void RB_tree::right_rotate(node *p)
{
     if(p->left==NULL)
          return ;
     else
     {
         node *y=p->left;
         if(y->right!=NULL)
         {
                  p->left=y->right;
                  y->right->parent=p;
         }
         else
                 p->left=NULL;
         if(p->parent!=NULL)
                 y->parent=p->parent;
         if(p->parent==NULL)
               root=y;
         else
         {
             if(p==p->parent->left)
                   p->parent->left=y;
             else
                   p->parent->right=y;
         }
         y->right=p;
         p->parent=y;
     }
}
node* RB_tree::successor(node *p)
{
      node *y=NULL;
     if(p->left!=NULL)
     {
         y=p->left;
         while(y->right!=NULL)
              y=y->right;
     }
     else
     {
         y=p->right;
         while(y->left!=NULL)
              y=y->left;
     }
     return y;
}
void RB_tree::disp()
{
     display(root);
}
void RB_tree::display(node *p)
{
     if(root==NULL)
     {
          cout<<"\nEMP.";
          return ;
     }
     if(p!=NULL)
     {
                cout<<"\n\t NODE: ";
                cout<<"\n KEY: "<<p->key;
                cout<<"\n COLOR: "; if(p->color=='b')
     cout<<"BLACK";
    else
     cout<<"Red"; if(p->parent!=NULL)
                       cout<<"\n PARENT: "<<p->parent->key;
                else
                       cout<<"\n THERE IS NO PARENT OF THE NODE. "; if(p->right!=NULL)
                       cout<<"\n RIGHT CHILD: "<<p->right->key;
                else
                       cout<<"\n THERE IS NO RIGHT CHILD OF THE NODE.. "; if(p->left!=NULL)
                       cout<<"\n LEFT CHILD: "<<p->left->key;
                else
                       cout<<"\n THERE IS NO LEFT CHILD OF THE NODE.  ";
                cout<<endl; if(p->left)
    {
                 cout<<"\n\nLEFT:\n"; display(p->left);
    }

    if(p->right)
    {
         cout<<"\n\nRIGHT:\n"; display(p->right);
    }
     }
}
void RB_tree::search()
{
     if(root==NULL)
     {
           cout<<"\nEMPTY TREE\n" ;
           return  ;
     }
     int x;
     cout<<"\n ENTER THE KEY TO BE SEARCHED: "; cin>>x;
     node *p=root;
     int found=0;
     while(p!=NULL&& found==0)
     {
            if(p->key==x)
                found=1;
            if(found==0)
            {
                 if(p->key<x) p=p->right;
                 else
                      p=p->left;
            }
     }
     if(found==0)
          cout<<"\nELEMENT NOT FOUND.";
     else
     {
                cout<<"\n\t FOUND NODE: ";
                cout<<"\n KEY: "<<p->key;
                cout<<"\n COLOUR: "; if(p->color=='b')
     cout<<"BLACK";
    else
     cout<<"RED"; if(p->parent!=NULL)
                       cout<<"\n PARENT: "<<p->parent->key;
                else
                       cout<<"\n THERE IS NO PARENT OF THE NODE. "; if(p->right!=NULL)
                       cout<<"\n RIGHT CHILD: "<<p->right->key;
                else
                       cout<<"\n THERE IS NO RIGHT CHILD OF THE NODE. "; if(p->left!=NULL)
                       cout<<"\n LEFT CHILD: "<<p->left->key;
                else
                       cout<<"\n THERE IS NO LEFT CHILD OF THE NODE.  ";
                cout<<endl;
     }
}
int main()
{
    int ch,y=0;
    RB_tree obj;
    cout<<"\n\t--------RED BLACK TREE---------" ;
    do
    {               
                cout<<"\n 1. INSERT NODE";
                cout<<"\n 2. DELETE NODE";
                cout<<"\n 3. SEARCH FOR ELEMENT";
                cout<<"\n 4. DISPLAY ";
                cout<<"\n 5. EXIT " ;
                cout<<"\nSelect the operation you want to perform : "; cin>>ch;
                switch(ch)
                {
                          case 1 : obj.insert();
                                   cout<<"\nNODE INSERTED.\n";
                                   break;
                          case 2 : obj.del();
                                   break;
                          case 3 : obj.search();
                                   break;
                          case 4 : obj.disp();
                                   break;
                          case 5 : y=1;
                                   break;
                          default : cout<<"\nENTER A VALID CHOICE!!.";
                }
                cout<<endl;
    }while(y!=1);
    return 1;
}

File: Python/Trees/red_black.py
# Define Node
class Node():
    def __init__(self,val):
        self.val = val                           
        self.parent = None                     
        self.left = None                             
        self.right = None                              
        self.color = 1                                 

# Define R-B Tree
class RBTree():
    def __init__(self):
        self.NULL = Node ( 0 )
        self.NULL.color = 0
        self.NULL.left = None
        self.NULL.right = None
        self.root = self.NULL


    # Insert New Node
    def insertNode(self, key):
        node = Node(key)
        node.parent = None
        node.val = key
        node.left = self.NULL
        node.right = self.NULL
        node.color = 1                                

        y = None
        x = self.root

        while x != self.NULL :                         
            y = x
            if node.val < x.val :
                x = x.left
            else :
                x = x.right

        node.parent = y                          
        if y == None :                         
            self.root = node
        elif node.val < y.val :            
            y.left = node
        else :
            y.right = node

        if node.parent == None :                     
            node.color = 0
            return

        if node.parent.parent == None :              
            return

        self.fixInsert ( node )                      

    def minimum(self, node):
        while node.left != self.NULL:
            node = node.left
        return node


    # Code for left rotate
    def LR ( self , x ) :
        y = x.right                               
        x.right = y.left                            
        if y.left != self.NULL :
            y.left.parent = x

        y.parent = x.parent                   
        if x.parent == None :                   
            self.root = y                              
        elif x == x.parent.left :
            x.parent.left = y
        else :
            x.parent.right = y
        y.left = x
        x.parent = y


    # Code for right rotate
    def RR ( self , x ) :
        y = x.left                                  
        x.left = y.right                              
        if y.right != self.NULL :
            y.right.parent = x

        y.parent = x.parent                         
        if x.parent == None :                      
            self.root = y                         
        elif x == x.parent.right :
            x.parent.right = y
        else :
            x.parent.left = y
        y.right = x
        x.parent = y


    # Fix Up Insertion
    def fixInsert(self, k):
        while k.parent.color == 1:                     
            if k.parent == k.parent.parent.right:       
                u = k.parent.parent.left              
                if u.color == 1:                      
                    u.color = 0                      
                    k.parent.color = 0
                    k.parent.parent.color = 1       
                    k = k.parent.parent              
                else:
                    if k == k.parent.left:         
                        k = k.parent
                        self.RR(k)                       
                    k.parent.color = 0
                    k.parent.parent.color = 1
                    self.LR(k.parent.parent)
            else:                                   
                u = k.parent.parent.right             
                if u.color == 1:                      
                    u.color = 0                          
                    k.parent.color = 0
                    k.parent.parent.color = 1           
                    k = k.parent.parent                 
                else:
                    if k == k.parent.right:             
                        k = k.parent
                        self.LR(k)                        
                    k.parent.color = 0
                    k.parent.parent.color = 1
                    self.RR(k.parent.parent)              
            if k == self.root:                           
                break
        self.root.color = 0                               


    # fixing problems after deletion
    def fixDelete ( self , x ) :
        while x != self.root and x.color == 0 :           
            if x == x.parent.left :                      
                s = x.parent.right                       
                if s.color == 1 :                         
                    s.color = 0                           
                    x.parent.color = 1                    
                    self.LR ( x.parent )                  
                    s = x.parent.right
                # If both the child are black
                if s.left.color == 0 and s.right.color == 0 :
                    s.color = 1                           
                    x = x.parent
                else :
                    if s.right.color == 0 :               
                        s.left.color = 0                  
                        s.color = 1                      
                        self.RR ( s )                    
                        s = x.parent.right

                    s.color = x.parent.color
                    x.parent.color = 0                   
                    s.right.color = 0
                    self.LR ( x.parent )                 
                    x = self.root
            else :                                       
                s = x.parent.left                         
                if s.color == 1 :                         
                    s.color = 0                           
                    x.parent.color = 1                    
                    self.RR ( x.parent )                 
                    s = x.parent.left

                if s.right.color == 0 and s.right.color == 0 :
                    s.color = 1
                    x = x.parent
                else :
                    if s.left.color == 0 :                
                        s.right.color = 0                
                        s.color = 1
                        self.LR ( s )                     
                        s = x.parent.left

                    s.color = x.parent.color
                    x.parent.color = 0
                    s.left.color = 0
                    self.RR ( x.parent )
                    x = self.root
        x.color = 0


    # Function to transplant nodes
    def __rb_transplant ( self , u , v ) :
        if u.parent == None :
            self.root = v
        elif u == u.parent.left :
            u.parent.left = v
        else :
            u.parent.right = v
        v.parent = u.parent


    # Function for deletion
    def delete_node_helper ( self , node , key ) :
        z = self.NULL
        while node != self.NULL :                         
            if node.val == key :
                z = node

            if node.val <= key :
                node = node.right
            else :
                node = node.left

        if z == self.NULL :                                
            print ( "VALUE NOT PRESENT IN TREE !!" )
            return

        y = z
        y_original_color = y.color                          
        if z.left == self.NULL :                            
            x = z.right                                     
            self.__rb_transplant ( z , z.right )           
        elif (z.right == self.NULL) :                       
            x = z.left                                      
            self.__rb_transplant ( z , z.left )             
        else :                                              
            y = self.minimum ( z.right )                    
            y_original_color = y.color                      
            x = y.right
            if y.parent == z :                              
                x.parent = y                                
            else :
                self.__rb_transplant ( y , y.right )
                y.right = z.right
                y.right.parent = y

            self.__rb_transplant ( z , y )
            y.left = z.left
            y.left.parent = y
            y.color = z.color
        if y_original_color == 0 :                         
            self.fixDelete ( x )


    # Deletion of node
    def delete_node ( self , val ) :
        self.delete_node_helper ( self.root , val )         


    # Function to print
    def __printCall ( self , node , indent , last ) :
        if node != self.NULL :
            print(indent, end=' ')
            if last :
                print ("R----",end= ' ')
                indent += "     "
            else :
                print("L----",end=' ')
                indent += "|    "

            s_color = "RED" if node.color == 1 else "BLACK"
            print ( str ( node.val ) + "(" + s_color + ")" )
            self.__printCall ( node.left , indent , False )
            self.__printCall ( node.right , indent , True )

    # Function to call print
    def print_tree ( self ) :
        self.__printCall ( self.root , "" , True )



tree = RBTree()
print("Select the operation you want to perform: ")
print("1. INSERT NODE")
print("2. DELETE NODE")
print("3. DISPLAY")
print("4. EXIT")

exit = True
while exit:
    
    choice = int(input("ENTER THE OPERATION NUMBER YOU WANT TO PERFORM: "))
    
    if(choice == 1):
        data = int(input("ENTER THE DATA YOU WANT TO INSERT INTO TREE: "))
        tree.insertNode(data)
    elif(choice == 2):
        val = int(input("ENTER THE NODE VALUE WHICH YOU WANT TO DELETE: "))
        tree.delete_node(val)
    elif(choice == 3):
        tree.print_tree()
    elif(choice == 4):
        exit = False
    else:
        print("INVALID INPUT!!!")

